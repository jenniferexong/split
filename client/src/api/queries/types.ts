import { ApiProduct, ApiReceipt, ApiPerson, ApiStore } from 'api';

export type QueryResult<T> = () => Promise<T>;

export interface ProductResult {
  product: ApiProduct;
}

export interface ProductsResult {
  products: ApiProduct[];
}

export interface StoreResult {
  store: ApiStore;
}

export interface StoresResult {
  stores: ApiStore[];
}

export interface PeopleResult {
  people: ApiPerson[];
}

export interface PersonResult {
  person: ApiPerson;
}

export interface ReceiptResult {
  receipt: ApiReceipt;
}

export interface DeleteReceiptResult {
  deleteReceipt: ApiReceipt;
}

export interface ReceiptsResult {
  receipts: ApiReceipt[];
}
