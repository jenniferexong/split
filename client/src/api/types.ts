export interface ApiProduct {
  id: number;
  name: string;
}

export interface ApiStore {
  id: number;
  name: string;
}

export interface ApiPerson {
  id: number;
  firstName: string;
  lastName: string;
  email: string;
}

// TODO
export interface ApiReceipt {
  id: number;
}
