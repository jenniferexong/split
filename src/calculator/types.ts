export interface Person {
  name: string;
}

export interface EntryType {
  title?: string;
  /** Should default to mine if not specified */
  whose: "mine" | "theirs" | "split";
  price: number;
}

export interface ReceiptType {
  title: string;
  entries: EntryType[];
  subtotal: number;
}
