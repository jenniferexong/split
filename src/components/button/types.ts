import { ReactNode } from 'react';

export interface ButtonProps {
  /** Button text */
  children: ReactNode;
  onClick(): void;
}
