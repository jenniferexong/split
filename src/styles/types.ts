export type Theme = {
  background: string;

  baseFont: {
    font: string;
    size: string;
    color: string;
    lineHeight: string;
  };

  colors: {
    accent: string;
  };

  components: {
    button: {
      background: string;
      textColor: string;
    };
  };
};
