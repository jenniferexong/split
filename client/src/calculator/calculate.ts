import { unreachable } from '../utils/unreachable';
import { AppType, InvoiceData } from './types';

interface CalculateResult {
  globalTotal: number;
  /** In the same order of people in the app */
  invoices: InvoiceData[];
}

export const calculate = (data: AppType): CalculateResult => {
  let globalTotal = 0;

  const invoices: InvoiceData[] = [
    {
      totalSpendings: 0,
      actualSpendings: 0,
      oweings: 0,
    },
    {
      totalSpendings: 0,
      actualSpendings: 0,
      oweings: 0,
    },
  ];

  data.people.forEach((person, personIndex) => {
    person.receipts.forEach(receipt => {
      receipt.items.forEach(item => {
        const { price, whose } = item;

        globalTotal += price;
        invoices[personIndex].actualSpendings += price;

        switch (whose) {
          case 'mine': {
            invoices[personIndex].totalSpendings += price;
            break;
          }
          case 'theirs': {
            invoices[personIndex ^ 1].totalSpendings += price;
            invoices[personIndex ^ 1].oweings += price;
            invoices[personIndex].oweings -= price;
            break;
          }
          case 'split': {
            const splitCost = price / 2;
            invoices[personIndex].totalSpendings += splitCost;
            invoices[personIndex ^ 1].totalSpendings += splitCost;
            invoices[personIndex ^ 1].oweings += splitCost;
            invoices[personIndex].oweings -= splitCost;
            break;
          }
          default: {
            unreachable(whose);
          }
        }
      });
    });
  });

  invoices.forEach(invoice => {
    invoice.oweings = invoice.oweings < 0 ? 0 : invoice.oweings;
  });

  return { globalTotal, invoices };
};
