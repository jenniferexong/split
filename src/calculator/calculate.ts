import { AppType, PersonType } from "./types";

/**
 * Calculates who owes who, and how much is owed.
 *
 * NOTE only works with two people
 */
export const calculateOwings = (
  data: AppType
): { ower: string; owee: string; amount: number } => {
  if (data.people.length !== 2) throw new Error("Two people required");

  const person1 = {
    name: data.people[0].name,
    owings: calculateOwing(data.people[1]),
  };

  const person2 = {
    name: data.people[1].name,
    owings: calculateOwing(data.people[0]),
  };

  let ower;
  let owee;
  if (person1.owings > person2.owings) {
    ower = person1;
    owee = person2;
  } else {
    ower = person2;
    owee = person1;
  }

  return {
    ower: ower.name,
    owee: owee.name,
    amount: ower.owings - owee.owings,
  };
};

/**
 * Calculates the amount the other person owes to this person
 */
const calculateOwing = (person: PersonType): number => {
  let owings = 0;
  for (let receipt of person.receipts) {
    for (let item of receipt.items) {
      if (item.whose === "split") {
        owings += item.price / 2;
      } else if (item.whose === "theirs") {
        owings += item.price;
      }
    }
  }
  return owings;
};

export const calculateSpendings = ({
  people,
}: AppType): { name: string; spendings: number }[] => {
  const spendings1 = calculateSpending(people[0]);
  const spendings2 = calculateSpending(people[1]);

  return [
    {
      name: people[0].name,
      spendings: spendings1.mySpendings + spendings2.theirSpendings,
    },
    {
      name: people[1].name,
      spendings: spendings2.mySpendings + spendings1.theirSpendings,
    },
  ];
};

/**
 * Calculates the total spendings made by each person,
 * for receipts paid by a given person
 */
const calculateSpending = (
  person: PersonType
): { mySpendings: number; theirSpendings: number } => {
  let mySpendings = 0;
  let theirSpendings = 0;
  for (let receipt of person.receipts) {
    for (let { whose, price } of receipt.items) {
      if (whose === "mine") {
        mySpendings += price;
      } else if (whose === "split") {
        mySpendings += price / 2;
        theirSpendings += price / 2;
      } else {
        theirSpendings += price;
      }
    }
  }
  return { mySpendings, theirSpendings };
};
