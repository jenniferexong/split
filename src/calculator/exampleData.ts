import { AppType, ReceiptType } from './types';

export const exampleReceipt: ReceiptType = {
  title: 'Countdown',
  items: [
    { title: 'Pork', whose: 'split', price: 4.56 },
    {
      title: 'Carrots',
      whose: 'split',
      price: 2.69,
    },
    {
      title: 'Apples',
      whose: 'mine',
      price: 1.5,
    },
    {
      title: '',
      whose: 'theirs',
      price: 10.4,
    },
  ],
  subtotal: 19.15,
};

export const testData: AppType = {
  people: [
    {
      name: 'Jennifer',
      receipts: [
        {
          title: 'Seasons Market',
          items: [
            {
              title: 'cucumbers',
              whose: 'split',
              price: 5,
            },
            {
              title: 'spring onion',
              whose: 'split',
              price: 1.99,
            },
          ],
          subtotal: 6.99,
        },
        {
          title: 'Countdown',
          items: [
            {
              title: 'berries',
              whose: 'mine',
              price: 18,
            },
            {
              title: 'strawberies',
              whose: 'theirs',
              price: 7,
            },
            {
              title: 'rubbish bags',
              whose: 'split',
              price: 5.49,
            },
            {
              title: 'butter',
              whose: 'mine',
              price: 8.5,
            },
          ],
          subtotal: 38.99,
        },
      ],
    },
    {
      name: 'Andy',
      receipts: [
        {
          title: 'Greens',
          items: [
            {
              title: 'choy sum',
              whose: 'split',
              price: 11.96,
            },
          ],
          subtotal: 11.96,
        },
        {
          title: 'Countdown',
          items: [
            {
              title: 'mushrooms',
              whose: 'mine',
              price: 4.24,
            },
            {
              title: 'chicken',
              whose: 'split',
              price: 16,
            },
            {
              title: 'bacon',
              whose: 'split',
              price: 5,
            },
            {
              title: 'flour',
              whose: 'split',
              price: 2,
            },
            {
              title: 'cheese',
              whose: 'split',
              price: 11,
            },
            {
              title: 'pork loin',
              whose: 'split',
              price: 15,
            },
            {
              title: 'chicken stock',
              whose: 'split',
              price: 2.5,
            },
            {
              title: 'sponges',
              whose: 'split',
              price: 4,
            },
            {
              title: 'capsicum',
              whose: 'split',
              price: 2.99,
            },
            {
              title: 'all spice',
              whose: 'split',
              price: 2.1,
            },
            {
              title: 'rice',
              whose: 'split',
              price: 7,
            },
            {
              title: 'eggs',
              whose: 'split',
              price: 11.8,
            },
          ],
          subtotal: 83.63,
        },
        {
          title: 'Tai Ping',
          items: [
            {
              title: 'buns',
              whose: 'split',
              price: 20.97,
            },
            {
              title: 'cucumber',
              whose: 'split',
              price: 1.99,
            },
            {
              title: 'mushrooms',
              whose: 'split',
              price: 12.99,
            },
            {
              title: 'carrots',
              whose: 'split',
              price: 1.24,
            },
            {
              title: 'lgm black bean',
              whose: 'mine',
              price: 10.99,
            },
            {
              title: 'lgm pork',
              whose: 'theirs',
              price: 6.99,
            },
            {
              title: 'hotpot seasoning',
              whose: 'split',
              price: 5.39,
            },
            {
              title: 'lamb slices',
              whose: 'split',
              price: 27.98,
            },
            {
              title: 'fish balls',
              whose: 'split',
              price: 3.99,
            },
            {
              title: 'fish tofu',
              whose: 'split',
              price: 7,
            },
            {
              title: 'tripe',
              whose: 'mine',
              price: 13.98,
            },
            {
              title: 'spring onion',
              whose: 'split',
              price: 1.99,
            },
          ],
          subtotal: 115.5,
        },
      ],
    },
  ],
};
