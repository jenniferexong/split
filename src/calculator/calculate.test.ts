import { calculateOwings } from './calculate';
import { exampleReceipt } from './exampleData';
import { AppType } from './types';

describe('calculate', () => {
  const data: AppType = {
    people: [
      {
        name: 'Person 1',
        receipts: [
          {
            title: 'Countdown',
            items: [
              { title: '', whose: 'split', price: 550 },
              { title: '', whose: 'split', price: 440 },
              { title: '', whose: 'mine', price: 20.5 },
              { title: '', whose: 'theirs', price: 21.8 },
            ],
            // we don't care about subtotal for this
            subtotal: 0,
          },
        ],
      },
      {
        name: 'Person 2',
        receipts: [
          {
            title: 'New World',
            items: [
              { title: '', whose: 'mine', price: 4.56 },
              { title: '', whose: 'theirs', price: 2.69 },
              { title: '', whose: 'mine', price: 1.5 },
              { title: '', whose: 'mine', price: 10.4 },
              { title: '', whose: 'theirs', price: 1.4 },
              { title: '', whose: 'split', price: 30.5 },
            ],
            // we don't care about subtotal for this
            subtotal: 0,
          },
        ],
      },
    ],
  };

  describe('calculates the correct information', () => {
    test('case 1', () => {
      const { ower, owee, amount } = calculateOwings(data);
      expect(ower).toBe('Person 2');
      expect(owee).toBe('Person 1');
      expect(amount).toBe(497.46);
    });

    test('case 2', () => {
      const exampleData = {
        people: [
          {
            name: 'Person 1',
            receipts: [exampleReceipt],
          },
          {
            name: 'Person 2',
            receipts: [exampleReceipt],
          },
        ],
      };
      expect(calculateOwings(exampleData).amount).toBe(0);
    });
  });
});
