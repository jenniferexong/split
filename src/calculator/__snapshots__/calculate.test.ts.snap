// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate calculates the correct information case 1 1`] = `
Object {
  "globalTotal": 1083.3500000000001,
  "invoices": Array [
    Object {
      "actualSpendings": 1032.3,
      "oweings": 0,
      "totalSpendings": 534.84,
    },
    Object {
      "actualSpendings": 51.05,
      "oweings": 497.4599999999999,
      "totalSpendings": 548.5099999999999,
    },
  ],
}
`;

exports[`calculate calculates the correct information case 2 1`] = `
Object {
  "globalTotal": 38.3,
  "invoices": Array [
    Object {
      "actualSpendings": 19.15,
      "oweings": 0,
      "totalSpendings": 19.15,
    },
    Object {
      "actualSpendings": 19.15,
      "oweings": 0,
      "totalSpendings": 19.15,
    },
  ],
}
`;
